SNOW-778564: Executing large batch for SQL updating > max integer rows fails with message "value is too large to be stored as integer".

<!--
If you need urgent assistance then file the issue using the support process: 
https://community.snowflake.com/s/article/How-To-Submit-a-Support-Case-in-Snowflake-Lodge 
otherwise continue here. 
-->


Please answer these questions before submitting your issue. 
In order to accurately debug the issue this information is required. Thanks!

1. What version of JDBC driver are you using?
3.13.25
   
2. What operating system and processor architecture are you using?
macOS 12.5.1 (M1)
   
3. What version of Java are you using?
OpenJDK 11.0.12
   
4. What did you do?

I attempted to execute a Snowflake SQL query as a batch that would update a number of rows greater than the max integer value, necessitating a call to `executeLargeBatch()`. I added a Snowflake SQL statement, which I expected to update 2.2 billion rows, to a java.sql.PreparedStatement, added a batch to the PreparedStatement, and then attempted to call `executeLargeBatch()` on the PreparedStatement to execute it.

E.g.
```
// SQL expected to update > max integer rows
String sqlStatement = "UPDATE TEST_TABLE SET COL1 = 'TESTING';

PreparedStatement preparedStatement = connection.prepareStatement(sqlStatement);
preparedStatement.addBatch();
long[] queryResult = preparedStatement.executeLargeBatch();
```

5. What did you expect to see?

My expectation was that `preparedStatement.executeLargeBatch()` would return a `long[]` containing one entry, an updated row count value of 2.2 billion.

Instead I observed a SnowflakeSQLLoggedException being thrown with message "Value is too large to be stored as integer at batch index 0. Use executeLargeBatch() instead."
```
java.sql.BatchUpdateException: Value is too large to be stored as integer at batch index 0. Use executeLargeBatch() instead.
    at net.snowflake.client.jdbc.SnowflakeStatementV1.executeBatchInternal(SnowflakeStatementV1.java:460)
    at net.snowflake.client.jdbc.SnowflakePreparedStatementV1.executeBatchInternalWithArrayBind(SnowflakePreparedStatementV1.java:857)
    at net.snowflake.client.jdbc.SnowflakePreparedStatementV1.executeLargeBatch(SnowflakePreparedStatementV1.java:796)
    at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeLargeBatch(HikariProxyPreparedStatement.java)
    ...
Caused by: net.snowflake.client.jdbc.SnowflakeSQLLoggedException: Value is too large to be stored as integer at batch index 0. Use executeLargeBatch() instead.
    at net.snowflake.client.jdbc.SnowflakeStatementV1.executeBatchInternal(SnowflakeStatementV1.java:436)
    ... 29 more
```

--- a/src/main/java/net/snowflake/client/jdbc/SnowflakePreparedStatementV1.java
+++ b/src/main/java/net/snowflake/client/jdbc/SnowflakePreparedStatementV1.java
@@ -877,8 +877,7 @@ class SnowflakePreparedStatementV1 extends SnowflakeStatementV1
       } else {
         // Array binding is not supported
         if (isLong) {
-          updateCounts.longArr = executeBatchInternal(false).longArr;
-
+          updateCounts.longArr = executeBatchInternal(true).longArr;
         } else {
           updateCounts.intArr = executeBatchInternal(false).intArr;
         }

--- a/src/test/java/net/snowflake/client/jdbc/PreparedStatement2LatestIT.java
+++ b/src/test/java/net/snowflake/client/jdbc/PreparedStatement2LatestIT.java
@@ -317,4 +317,30 @@ public class PreparedStatement2LatestIT extends PreparedStatement0IT {
     assertTrue(rs.next());
     assertTrue(prepStatement.unwrap(SnowflakePreparedStatementV1.class).isAlreadyDescribed());
   }
+
+  @Test
+  public void testExecuteLargeBatchOverMaxInt() throws SQLException {
+    Connection connection = null;
+    try {
+      connection = getConnection();
+      try (Statement statement = connection.createStatement()) {
+        statement.execute("create or replace table over_int(id decimal);");
+        for (int i = 1; i <= 11; i++) {
+          statement.execute(
+              "insert into over_int (SELECT seq8() FROM table(generator(rowCount => 200000000)));");
+        }
+        String sqlStatement = "UPDATE over_int SET id = 200";
+
+        PreparedStatement pstmt = connection.prepareStatement(sqlStatement);
+        pstmt.addBatch();
+        long[] queryResult = pstmt.executeLargeBatch();
+        assertEquals(1, queryResult.length);
+      }
+    } finally {
+      if (connection != null) {
+        connection.createStatement().execute("drop table if exists over_int ");
+        connection.close();
+      }
+    }
+  }
 }

